File: project_euler/Problem016/Problem016.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef struct {
    int *value;
    int length;
} Decimal;

static void get_power(int value, Decimal *decimal, int *auxiliary, int length);
static void get_square(Decimal *decimal, int *auxiliary, int length);
static void get_double(Decimal *decimal);

void Problem016(void) {
    int target = 1000;
    int length = (target + 4) / 5;
    Decimal *decimal = (Decimal *) malloc(sizeof(Decimal));
    decimal->value = (int *) malloc(sizeof(int) * length);
    for (int i = 0; i < length; ++i) {
        decimal->value[i] = 0;
    }
    decimal->length = 0;
    int *auxiliary = (int *) malloc(sizeof(int) * length);
    get_power(target, decimal, auxiliary, length);

    int result = 0;
    for (int i = 0; i < decimal->length; ++i) {
        int value = decimal->value[i];
        while (value > 0) {
            result += value % 10;
            value /= 10;
        }
    }
    printf("%d\n", result);
}

static void get_power(int value, Decimal *decimal, int *auxiliary, int length) {
    if (value < 8) {
        decimal->value[0] = (int) pow(2, value);
        decimal->length = 1;
        return;
    }

    get_power(value >> 1, decimal, auxiliary, length);
    get_square(decimal, auxiliary, length);
    if (value & 1) {
        get_double(decimal);
    }
}

static void get_square(Decimal *decimal, int *auxiliary, int length) {
    for (int i = 0; i < length; ++i) {
        auxiliary[i] = 0;
    }
    for (int i = 0; i < decimal->length; ++i) {
        for (int j = 0; j < decimal->length; ++j) {
            long long result = (long long) decimal->value[i] * decimal->value[j];
            int index = i + j;
            auxiliary[index] += (int) (result % (int) 1e5);
            auxiliary[index + 1] += (int) (result / (int) 1e5);
        }
    }

    decimal->length = (decimal->length << 1) - 1;
    for (int i = 0; i < decimal->length; ++i) {
        if (auxiliary[i] >= (int) 1e5) {
            auxiliary[i + 1] += auxiliary[i] / (int) 1e5;
            auxiliary[i] %= (int) 1e5;
        }
    }
    if (auxiliary[decimal->length]) {
        ++decimal->length;
    }
    memcpy(decimal->value, auxiliary, sizeof(int) * decimal->length);
}

static void get_double(Decimal *decimal) {
    for (int i = 0; i < decimal->length; ++i) {
        decimal->value[i] <<= 1;
    }

    for (int i = 0; i < decimal->length; ++i) {
        if (decimal->value[i] >= (int) 1e5) {
            decimal->value[i + 1] += decimal->value[i] / (int) 1e5;
            decimal->value[i] %= (int) 1e5;
        }
    }
    if (decimal->value[decimal->length]) {
        ++decimal->length;
    }
}

File: project_euler/Problem016/Problem016.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem016() {
    target := 1000
    length := (target + 4) / 5
    decimal := Decimal{make([]int, length), 0}
    for i := 0; i < length; i++ {
        decimal.value[i] = 0
    }
    auxiliary := make([]int, length)
    getPower(target, &decimal, auxiliary, length)

    result := 0
    for _, value := range decimal.value[:decimal.length] {
        for value > 0 {
            result += value % 10
            value /= 10
        }
    }
    fmt.Println(result)
}

func getPower(value int, decimal *Decimal, auxiliary []int, length int) {
    if value < 8 {
        decimal.value[0] = int(math.Pow(2, float64(value)))
        decimal.length = 1
        return
    }

    getPower(value >> 1, decimal, auxiliary, length)
    getSquare(decimal, auxiliary, length)
    if 0 != (value & 1) {
        getDouble(decimal)
    }
}

func getSquare(decimal *Decimal, auxiliary []int, length int) {
    for i := 0; i < length; i++ {
        auxiliary[i] = 0
    }
    for i := 0; i < decimal.length; i++ {
        for j := 0; j < decimal.length; j++ {
            result := int64(decimal.value[i] * decimal.value[j])
            index := i + j
            auxiliary[index] += int(result) % int(1e5)
            auxiliary[index + 1] += int(result) / int(1e5)
        }
    }

    decimal.length = (decimal.length << 1) - 1
    for i := 0; i < decimal.length; i++ {
        if auxiliary[i] >= int(1e5) {
            auxiliary[i + 1] += auxiliary[i] / int(1e5)
            auxiliary[i] %= int(1e5)
        }
    }
    if 0 != auxiliary[decimal.length] {
        decimal.length++
    }
    copy(decimal.value, auxiliary)
}

func getDouble(decimal *Decimal)  {
    for i := 0; i < decimal.length; i++ {
        decimal.value[i] <<= 1
    }

    for i := 0; i < decimal.length; i++ {
        if decimal.value[i] >= int(1e5) {
            decimal.value[i + 1] += decimal.value[i] / int(1e5)
            decimal.value[i] %= int(1e5)
        }
    }
    if 0 != decimal.value[decimal.length] {
        decimal.length++
    }
}

type Decimal struct {
    value []int
    length int
}
